src/buffer.c

```diff
@@ -190,30 +216,67 @@ buffer_write(struct Buffer *buffer, int fd) {
 size_t
 buffer_coalesce(struct Buffer *buffer, const void **dst) {
     size_t buffer_tail = (buffer->head + buffer->len) & buffer->size_mask;
+    size_t head = buffer->head;
 
     if (buffer_tail <= buffer->head) {
         /* buffer not wrapped */
+        if (buffer->type == SOCK_DGRAM)
+            head += sizeof(uint16_t);
         if (dst != NULL)
-            *dst = &buffer->buffer[buffer->head];
+            *dst = &buffer->buffer[head];
 
         return buffer->len;
     } else {
-        /* buffer wrapped */
-        size_t len = buffer->len;
-        char *temp = malloc(len);
-        if (temp != NULL) {
-            buffer_pop(buffer, temp, len);
-            assert(buffer->len == 0);
-
-            buffer_push(buffer, temp, len);
-            assert(buffer->head == 0);
-            assert(buffer->len == len);
-
-            free(temp);
+        if (buffer->type == SOCK_STREAM) {
+            /* buffer wrapped */
+            size_t len = buffer->len;
+            char *temp = malloc(len);
+            if (temp != NULL) {
+                buffer_pop(buffer, temp, len);
+                assert(buffer->len == 0);
+
+                buffer_push(buffer, temp, len);
+                assert(buffer->head == 0);
+                assert(buffer->len == len);
+
+                free(temp);
+            }
+        } else { /* SOCK_DGRAM */
+            /* buffer wrapped */
+            size_t len = buffer->len;
+            char *temp = malloc(len);
+            size_t bytes = 0, total = 0;
+            struct Buffer *newbuf = new_buffer(SOCK_DGRAM, len, EV_DEFAULT);
```

Comment:
It doesn't seem like we need allocate both `temp` and `newbuf`, how about always using a second buffer and a buffer to buffer move routine? I can take a stab at this later.

Reply:
OK, I'll defer to you doing that, it would help this for sure! Thanks for the offer of help!